File: providers/openrouter.py
"""OpenRouter provider (default). Free tier with many models."""

from __future__ import annotations

import json
import logging
from typing import AsyncIterator

import httpx

from .generic import GenericOpenAIProvider, _friendly_error

log = logging.getLogger(__name__)

OPENROUTER_BASE = "https://openrouter.ai/api/v1"
DEFAULT_MODEL = "google/gemma-3-27b-it"


class OpenRouterProvider(GenericOpenAIProvider):
    def __init__(self, config: dict) -> None:
        super().__init__(
            config,
            base_url=config.get("base_url", OPENROUTER_BASE),
        )
        self._model = config.get("model", DEFAULT_MODEL)

    def _chat_path(self) -> str:
        return "/chat/completions"

    def is_available(self) -> bool:
        return bool(self.config.get("api_key"))

    async def stream(self, prompt: str, system: str = "") -> AsyncIterator[str]:
        # OpenRouter requires HTTP-Referer for rate limiting attribution
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": "https://github.com/mariesqu/Quill",
            "X-Title": "Quill",
        }
        payload = {
            "model": self._model,
            "messages": messages,
            "stream": True,
        }

        async with httpx.AsyncClient(timeout=60) as client:
            async with client.stream(
                "POST",
                self._endpoint_url,
                json=payload,
                headers=headers,
            ) as response:
                if response.status_code >= 400:
                    body = await response.aread()
                    body_text = body.decode("utf-8", errors="replace")
                    log.error("OpenRouter API error %s: %s", response.status_code, body_text)
                    raise RuntimeError(
                        _friendly_error(response.status_code, body_text, self._model)
                    )

                async for line in response.aiter_lines():
                    if not line.startswith("data: "):
                        continue
                    data = line[6:]
                    if data.strip() == "[DONE]":
                        break
                    try:
                        chunk = json.loads(data)
                        choices = chunk.get("choices", [])
                        if not choices:
                            continue
                        delta = choices[0].get("delta", {})
                        content = delta.get("content", "")
                        if content:
                            yield content
                    except (json.JSONDecodeError, KeyError, IndexError, TypeError) as e:
                        log.debug("Failed to parse SSE chunk: %s — %s", data[:100], e)

File: ui/src-tauri/src/providers/openrouter.rs
use async_trait::async_trait;
use reqwest::Client;
use serde_json::json;

use crate::core::config::Config;
use super::{ChunkStream, Provider, friendly_error, openai_sse_stream};

pub struct OpenRouterProvider {
    client:  Client,
    model:   String,
    api_key: String,
    base_url: String,
}

impl OpenRouterProvider {
    pub fn new(cfg: &Config) -> Self {
        Self {
            client:  Client::new(),
            model:   cfg.model.clone(),
            api_key: cfg.api_key.clone().unwrap_or_default(),
            base_url: cfg.base_url.clone()
                .unwrap_or_else(|| "https://openrouter.ai/api/v1".into()),
        }
    }
}

#[async_trait]
impl Provider for OpenRouterProvider {
    async fn stream_completion(&self, system: &str, user: &str) -> Result<ChunkStream, String> {
        let url = format!("{}/chat/completions", self.base_url.trim_end_matches('/'));
        let body = json!({
            "model": self.model,
            "stream": true,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user",   "content": user},
            ]
        });

        let resp = self.client
            .post(&url)
            .bearer_auth(&self.api_key)
            .header("HTTP-Referer", "https://github.com/mariesqu/Quill")
            .header("X-Title", "Quill")
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Network error: {e}"))?;

        if !resp.status().is_success() {
            let status = resp.status().as_u16();
            let body = resp.text().await.unwrap_or_default();
            return Err(friendly_error(status, &body));
        }

        Ok(openai_sse_stream(resp))
    }
}
